### PR TITLE
[quant][pt2e] Support quantizer API in prepare_pt2e_quantizer

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97994

Summary:
This PR added a quantizer API to prepare_pt2e_quantizer, which enables user to annotate the nodes in the graph
directly to configure quantization, instead of relying on QConfigMapping, please see test cases in
test_quantize_pt2e.py for examples. Also added a prototype for QNNPackQuantizer, that will be modified later
to fully support different quantization capabilities of QNNPack/XNNPack

The goal for introducing quantizer is to add flexibility to the quantization API to allow modeling users and backend developers to express their quantization intentions programmably, which will free architecture optimization team from supporting different use cases in the core API in the future, as a concrete example, we used to have https://pytorch.org/docs/master/generated/torch.ao.quantization.qconfig_mapping.QConfigMapping.html#torch.ao.quantization.qconfig_mapping.QConfigMapping as the API for users to express their intent for quantization in fx graph mode quantization, and it has some fancy options like `set_module_name_regex` and `set_module_name_object_type_order`, this is not needed for all backends and adds burden of maintenance to AO team, in the quantizer API we will move these APIs to a backend specific `Quantizer` that needs this feature, and all the backends or even advanced modeling users can implement their own quantizer to express their intent for quantization through annotating the nodes, for example, to express the quantization intention of quantizing a convolution node, a user will find the convolution node in the graph and do:
```
operator_spec = qnnpack_quantizer.get_default_per_channel_symmetric_qnnpack_operator_spec()
conv_node.meta["target_dtype_info"] = {
    "input_act_obs_or_fq_ctr": _get_act_obs_or_fq_ctr(operator_spec),
    "weight_obs_or_fq_ctr": _get_weight_obs_or_fq_ctr(operator_spec)
    "bias_obs_or_fq_ctr": _get_bias_obs_or_fq_ctr(operator_spec),
    "output_act_obs_or_fq_ctr": _get_act_obs_or_fq_ctr(operator_spec),
    # TODO: validation of weight_index must be set if weight_obs_or_fq_ctr is set
    "weight_index": 1,
    # TODO: validation of bias_index must be set if bias_obs_or_fq_ctr is set
    "bias_index": 2,
}
```
each backend will introduce their own quantizer, e.g. QNNPackQuantizer, which may expose more convenient APIs for modeling users to configure the annotation, and different quantizer can compose with each other to annotate the graph correctly for quantization.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_simple_quantizer
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_conv

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/quantization/_pt2e/_quantizer.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/_quantizer.py:8]: we should not have backend config as user facing api at all

Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/_quantizer.py:8]: sure, what is the recommendation? are you talking about removing this from the argument and just have backend assign the member variable directly?

**Comment by supriyar** [comment]
> please see test cases in test_quantize_pt2e.py for examples

the test file doesn't seem to be included in this PR?

**Comment by jerryzh168** [comment]
> > please see test cases in test_quantize_pt2e.py for examples
> 
> the test file doesn't seem to be included in this PR?

yeah this PR is WIP, just put up earlier so people know what it will look like

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (kimishpatel) [inline torch/ao/quantization/quantizer/qnnpack_quantizer.py:22]: Why is this called tensorSpec. It should be QuantizationSpec.

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (kimishpatel) [inline torch/ao/quantization/quantizer/qnnpack_quantizer.py:63]: I would like to say this: quant supports nn.Conv2d: This means it supports nn.Conv2d but also funcitonal.conv2d. This is simpler to understand IMO and hides complexity. 

You can certainly argue that it is a bit confusing because when API says nn.Conv2d is supported, it is not saying it supports functional.conv2d, but all the modules and functional variants eventually map to functional variants. So I would like to keep it this way and maybe get more input to say if we need to enumerate functional variants here as well.

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (kimishpatel) [inline torch/ao/quantization/quantizer/qnnpack_quantizer.py:62]: I dont know why are these free methods. THis is exactly why I wanted them to be on Quantizer. I think its fine for this prototype but I am gonna add those to quantizer.

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (kimishpatel) [inline torch/ao/quantization/quantizer/qnnpack_quantizer.py:51]: Why OperatorSpec? I think QuantizationSpec is also good. Although given that this in quantization codebase, it does not have to be have Quantization in the name, but not sure what OperatorSpec means for user. For implementation it makes sense because you are talking about aten op.

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (kimishpatel) [inline torch/ao/quantization/quantizer/qnnpack_quantizer.py:109]: What is `operator_type`? Did you mean `obct_type`?

File: torch/ao/quantization/_pt2e/prepare.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/prepare.py:36]: We dont have to do it now but "target_dtype_info" does not seem to be related to quantization. Maybe just called it "quantization_spec"

File: torch/ao/quantization/_pt2e/prepare.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/prepare.py:45]: Dont you want to check if the input_activation is a valid obsever?

File: torch/ao/quantization/_pt2e/prepare.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/prepare.py:56]: What is the use of `named_module`?

File: torch/ao/quantization/_pt2e/prepare.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/prepare.py:60]: How would this work? I thought if you reuse `_maybe_insert_input_observers_for_node`, you will need a backend_config?

File: torch/ao/quantization/_pt2e/prepare.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/prepare.py:89]: Shouldnt all of this happen as part of `_maybe_insert_output_observer_for_node`? If not at least factor it out in a different function where you call `_maybe_insert_output_observer_for_node` and also update uses?

File: torch/ao/quantization/_pt2e/prepare.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/prepare.py:94]: `is_general_tensor_value_op` not defined.

File: torch/ao/quantization/_pt2e/prepare.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/prepare.py:105]: Why save_state?

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Left a few comments. Also move quantizer to _pt2 folder?

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:36]: yeah, we can change the name a bit later, I'll add a TODO

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:45]: you mean not None? it's probably handled in `_maybe_insert_input_observers_for_node` already

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (jerryzh168) [reply torch/ao/quantization/quantizer/qnnpack_quantizer.py:51]: yeah this is just something as a starting point, please feel free to modify to something makes more sense to qnnpack users, I think the current naming mainly reflects the granularity of the spec, one is for tensor, the other is for operator

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (jerryzh168) [reply torch/ao/quantization/quantizer/qnnpack_quantizer.py:62]: I think we can discuss, I'd like to start with qnnpack for now, we can decide before we advertise this to other backends if we want to put those in quantizer or not

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:56]: this is needed to store a mapping from observer name to observer instance right now: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/prepare.py#L663 maybe it can be improved later after refactoring

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:60]: yeah it turns out the only remaining thing we read from backend_config now is just the index of weight and bias, this PR removes it and puts them in the annotation dictionary instead: https://github.com/pytorch/pytorch/pull/98094

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:105]: convert code is expecting this, a lot of things are actually not needed for pt2e flow, but we can clean up later I think

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:94]: yeah will update

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:89]: makes sense, can I do this in a separate PR, I think we should be able to just move these things to `_maybe_insert_output_observer_for_node` and update all callsites

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:93]: probably don't need this, will remove

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (jerryzh168) [reply torch/ao/quantization/quantizer/qnnpack_quantizer.py:22]: this is initially called QuantizationParam, and OperatorSpec is called QuantizationSpec, feels a bit confusing though, but this file is supposed to be a starting point/prototype, feel free to modify since right now qnnpack/xnnpack is fully owning this

File: torch/ao/quantization/_pt2e/prepare.py
Comment (kimishpatel) [reply torch/ao/quantization/_pt2e/prepare.py:45]: No, I mean you wont call `_maybe_insert_input_observers_for_node` if `skip_inserting_observers`. But `skip_inserting_observers` only accounts for output_is_a_tensor.

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (kimishpatel) [reply torch/ao/quantization/quantizer/qnnpack_quantizer.py:51]: I dont think we should distinguish between qnnpack users or other backends. I dont want to block this PR but all these have to be in base quantizer

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (jerryzh168) [reply torch/ao/quantization/quantizer/qnnpack_quantizer.py:22]: can you postpone moving to base quantizer a bit, I would like to discuss with other people and gather more user data from backend developers before deciding

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:45]: oh I see, this logic is copied from fx graph mode quant code base, this is probably used to handle ops like "tensor.size()" etc. I'm not exactly sure what should be the rule for input activations but looks like it's not needed right now

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:105]: fx graph mode quantization has more features than pt2e right now, e.g. qat support, standalone_module, custom_module, it is also working with modules, a lot of states are for supporting these features

File: torch/ao/quantization/_pt2e/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/prepare.py:108]: similar to other states, convert code expects this right now, but I think we may be able to read from node.meta now if we have "nn_module_stack" or something stored there

File: torch/ao/quantization/quantizer/qnnpack_quantizer.py
Comment (kimishpatel) [reply torch/ao/quantization/quantizer/qnnpack_quantizer.py:22]: having annotate and validate as the only thing in base quantizer makes no sense to me. We at least have to enforce implementation of quantizer to expose, what quantization spec and modules/op they support.

Comment (kimishpatel) [reply torch/ao/quantization/quantizer/qnnpack_quantizer.py:22]: >  would like to discuss with other people and gather more user data from backend developers before deciding

FOr this can you concretely identify what the concerns are?

Comment (jerryzh168) [reply torch/ao/quantization/quantizer/qnnpack_quantizer.py]: clarified in chat, I'm fine with either as long as we are open to change according to data in the future

File: torch/ao/quantization/fx/prepare.py
Comment (vkuzo) [inline torch/ao/quantization/fx/prepare.py:596]: can we spell out why the production code path is changed?

I think for pt2e APIs, stamping to unblock is reasonable.  For changes in reusable code, it's good to minimize them as well as spell out what they are.

**Review by vkuzo — APPROVED** [review]

File: torch/ao/quantization/fx/prepare.py
Comment (jerryzh168) [reply torch/ao/quantization/fx/prepare.py:596]: sure, I can add a comment, this is for pt2e path since not all nodes is using this key right now

File: torch/ao/quantization/_quantize_pt2e.py
Comment (kimishpatel) [inline torch/ao/quantization/_quantize_pt2e.py:69]: THis is only valid for ptq?

File: torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:22]: nit: float16 is not something natively supported in executorch

File: torch/ao/quantization/_quantize_pt2e.py
Comment (andrewor14) [reply torch/ao/quantization/_quantize_pt2e.py:69]: Yes

File: torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:242]: Why do we need observer for every single node?

File: torch/ao/quantization/_pt2e/prepare.py
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/prepare.py:42]: If observer needs to be inserted for input, does this also do the appropriate graph rewrite?

File: test/quantization/fx/test_quantize_pt2e.py
Comment (kimishpatel) [inline test/quantization/fx/test_quantize_pt2e.py:180]: we should test for observer insertion as well?

**Pull request closed without merging**
